plugin/igus_plugin.cpp

```diff
@@ -0,0 +1,308 @@
+
+        // Get the joint pointers
+        joints = model->GetJoints();
+
+        // Set up the update event
+        update_connection = event::Events::ConnectWorldUpdateBegin(std::bind(&NUbotsIgusPlugin::update_robot, this));
+
+        get_reactor().on<Network<message::motion::ServoTargets>>().then(
+            [this](const message::motion::ServoTargets& targets) {
+                // Store the incoming commands in the queue for the next simulation frame
+                std::lock_guard<std::mutex> lock(command_mutex);
+                for (const auto& c : targets.targets()) {
+                    // Convert ServoID to JointID
+                    message::motion::ServoTarget joint_target(c);
+                    joint_target.set_id(servo_id_to_joint[c.id()]);
+                    joint_target.set_position(c.position() - joint_offsets[c.id()]);
+                    command_queue.push_back(joint_target);
+                }
+            });
+
+        // Setup a P-controller, with a gain of 4.
+        pid = common::PID(40.0, 0.0, 0.0);
+
+        // Apply the P-controller to the joint.
+        for (int i = 0; i < joints.size(); i++) {
+            model->GetJointController()->SetPositionPID(joints[i]->GetScopedName(), pid);
+
+            // This will set the initial positions of the robot
+            model->GetJointController()->SetPositionTarget(joints[i]->GetScopedName(), initial_positions[i]);
+
+            joints[i]->SetPosition(0, initial_positions[i]);
+
+            joints[i]->SetStopDissipation(0, 0.0);
+            joints[i]->SetStiffness(0, 3.25);
+            // joints[i]->SetStopStiffness (0, 0.125);
+            // joints[i]->Set
+            // joints[i]->SetStiffnessDamping(0, 4.0, 0.125, 0.0);
+            // joints[i]->Set
+            joints[i]->SetEffortLimit(0, 20.0);
+
+            joints[i]->SetStiffnessDamping(0, 3.25, 0.25, joints[i]->Position());
+        }
+    }
+
+private:
+    void apply_joint_command(const message::motion::ServoTarget& target) {
+        // Set the joint's gain by applying the
+        // P-controller to the joints for positions.
+        if (target.id() == JointID::L_ANKLE_ROLL || target.id() == JointID::R_ANKLE_ROLL) {
+            model->GetJointController()->SetPositionPID(joints[target.id()]->GetScopedName(),
+                                                        common::PID(target.gain() * 0.7));
+        }
+        else {
+            model->GetJointController()->SetPositionPID(joints[target.id()]->GetScopedName(),
+                                                        common::PID(target.gain() * 4.0));
+        }
+
+        // Get our seconds and nanos in c++ land
+        auto seconds = std::chrono::seconds(target.time().seconds());
+        auto nanos   = std::chrono::nanoseconds(target.time().nanos());
+
+        // Make a timestamp out of the summation of them
+        NUClear::clock::time_point t      = std::chrono::time_point<NUClear::clock>(seconds + nanos);
```

Comment:
This is totally wrong, there is no reason why the timebase should be the same.
This must be fixed before servo commands will work properly.
Most likely we need a wrapper message around these anyway to identify which model to control.

Reply:
I've added into the message the current time from the client, but I haven't yet used it for calculating velocities.